[RHCLOUD-18723] Cascade delete resources

MikelAlejoBR opened the pull request:
Implements deleting the following resources with cascade:

- Source
- Application
- Endpoint

The way it has been done is the following:

1. All the resources and sub resources are deleted from the database in a transaction. We use the "RETURNING" statement to get all the deleted resources.
2. We raise an event for every single resources we have deleted.
3. We collect all the authentications for every resource we have deleted.
4. We delete the authentications in batch and raise an event for every authentication deleted.

This has been done to not leave the client with dangling resources on the database. Not being able to delete authentications isn't that bad, since:

1. They are encrypted.
2. Nobody else will have a reference to them, so they will be unrecoverable from the app.

This situation, if it happens, will require manual action from our side. I assumed that it is better and easier to clean up dangling authentications than it is to clean dangling resources and sub resources.

---

As a collateral change: I refactored the tests so that we only use one connection for each test, instead of opening a new one. Until now every time a test was ran a new connection was being opened, which has ended up making Postgres complain about "too many active connections".

## Links

[[RHCLOUD-18723]](https://issues.redhat.com/browse/RHCLOUD-18723)

MikelAlejoBR committed:
feature: cascade delete for the resources

It moves the "vault path format" to a constant too so that we can
"freeze" what the Vault path should look like. This way any dependant
functions that expect that format —like "Create", "keysToMap" and
"searchKeys".

lindgrenj6 commented on service/subresource_destroyer.go:
```diff
@@ -0,0 +1,158 @@
+		var appIds []int64
+		for _, app := range applications {
+			appIds = append(appIds, app.ID)
+			err := RaiseEvent("Application.Delete", &app, headers)
```
One teeny thing - it has to be `<Resource>.delete`

MikelAlejoBR replied:
All right, fixed!

lindgrenj6 commented on service/subresource_destroyer.go:
```diff
@@ -0,0 +1,158 @@
+			authentications = append(authentications, auths...)
+		}
+
+	case "Application":
```
We do need to delete the `ApplicationAuthentications` as well - which might need to be deleted first due to fkeys.

MikelAlejoBR replied:
I cascade deleted the applications too, to be able to notify about the deleted application authentications as well.

MikelAlejoBR committed:
fix: "too many connections" error in tests

Due to the large number of database tests we've got, and how we openned
a new connection for each one of them, the database was starting to
complain about having too many active connections at once.

The test utils have been refactored to use a single connection for the
tests.

lindgrenj6 commented on service/subresource_destroyer.go:
```diff
@@ -0,0 +1,165 @@
+		}
+
+		authentications = append(authentications, auths...)
+	case "Endpoint":
```
Endpoints can have authentications attached to them - so we should probably add the cascade logic to that DAO as well.

MikelAlejoBR replied:
The enpoints' authentications already get cleaned up: we call `authenticationsDao.ListIdsForResource("Endpoint", []int64{resourceId})` and then we deal with those authentications outside the `switch`.

MikelAlejoBR committed:
fix: event suffix should be "destroy" and not "delete"

lpichler commented on service/subresource_destroyer.go:
```diff
@@ -0,0 +1,173 @@
+
+		// Raise an event for the deleted application authentications.
+		for _, appAuth := range applicationAuthentications {
+			err = RaiseEvent("ApplicationAuthentication.destroy", &appAuth, headers)
```
This is related to most likely with all raised events: 


```json
{
   "id":14,
   "created_at":"2022-04-11 17:28:38 UTC",
   "updated_at":"2022-04-11 17:28:38 UTC",
   "paused_at":null,
   "application_id":26,
   "authentication_id":22,
   "tenant":""
}
```

`tenant`  as an external tenant is missing in those payloads

MikelAlejoBR committed:
fix: return tenants with the associated objects

The tenants are required so that the deletion events are raised with the
"tenant" key populated.

lpichler reviewed (approved):
LGTM 👍
